### DISCUSSION TITLE
No results returned for query search

### DISCUSSION THREAD

**Discussion by rahul-raj** [ravendb/ravendb#18641] [Q&A] [answered]
Hello,

I inserted the below documents to RavenDB
![Screenshot 2024-06-09 212403](https://github.com/ravendb/ravendb/assets/517415/af48869f-18e2-4493-b2bd-76962ff637d9)

However, when I try to query for job title `software engineer`, I get zero results.
![Screenshot 2024-06-09 212933](https://github.com/ravendb/ravendb/assets/517415/6af1c63f-9197-4ef9-90c1-27aaa03b6dfa)

Here is the code that I have:
```
@GetMapping("/test")
    public ResponseEntity<List<Job>> get(@RequestParam("title") String title) {
        try(IDocumentSession session = store.openSession()) {
            List<Job> jobs = session.query(Job.class)
                    .whereEquals("jobTitle", title, false)
                    .toList();
            return ResponseEntity.ok(jobs);
        }

    }
```

And here is the `Job` class:
```
public class Job {
    @JsonProperty("job_title")
    private String jobTitle;
    @JsonProperty("location")
    private String location;

    public String getJobTitle() {
        return jobTitle;
    }

    public void setJobTitle(String jobTitle) {
        this.jobTitle = jobTitle;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }
}

```

Any idea what's going wrong here ?

Also, how to perform fuzzy search here? I followed  this: https://ravendb.net/docs/article-page/6.0/nodejs/client-api/session/querying/text-search/fuzzy-search
but I was getting error in console saying "fuzzy" search not supported.

**Comment by rahul-raj** [comment]
Update:
Got it fixed by correcting to:
`.whereEquals("job_title", title, false)`

However I want to know how to perform fuzzy search here? I followed this: https://ravendb.net/docs/article-page/6.0/nodejs/client-api/session/querying/text-search/fuzzy-search
but I was getting error in console saying "fuzzy" search not supported.

```
net.ravendb.client.exceptions.InvalidQueryException: Raven.Client.Exceptions.InvalidQueryException: Method 'Fuzzy' is not supported.
Query: from 'Jobs' where fuzzy(job_title = $p0, 0.5)
```

code:
```
try(IDocumentSession session = store.openSession()) {
            List<Job> jobs  = session.advanced().documentQuery(Job.class)
                    .whereEquals("job_title", title, false)
                    .fuzzy(0.5)
                    .toList();
            return ResponseEntity.ok(jobs);
        }

```

same issue with `proximity` as well:
```
try(IDocumentSession session = store.openSession()) {
            List<Job> jobs  = session
                    .advanced()
                    .documentQuery(Job.class)
                    .search("job_title", title)
                    .proximity(2)
                    .toList();
            return ResponseEntity.ok(jobs);
        }
```

My requirement is that I should be able to make a similarity search (similar job titles, spelling mistakes etc).
So, I'm not sure if it has to be similarity search or proximity search in the first place.

Also, I see that the below code gets me the closest of results that I wanted:
```
try(IDocumentSession session = store.openSession()) {
            List<Job> jobs  = session
                    .advanced()
                    .documentQuery(Job.class)
                    .search("job_title", title)
                    .toList();
            return ResponseEntity.ok(jobs);
        }
```

**Reply by ayende** [reply]
Can you call `toString()` on the query instead of `toList()` and post what you are seeing?
If there is an error, please post the _entire_ error including the stack trace.

**Reply by ayende** [reply, marked as answer]
Note that proximity means: `software backend developer` would match `proximity(search(job_title, ' "software developer" ' ), 2) `

FWIW, for proximity, you must use:

`.search("job_title", title).proximity(2)` - will only work after a `search`, not a `where`

For fuzzy, you must use: `where` before the `fuzzy`. 

But if you'll use: `search(job_title, 'software developer' )` if will work as well (note the lack of `"` there).

Basically, I wouldn't go with either fuzzy or proximity here. Fuzzy is meant to correct typos, mostly, but it is _much_ more expensive internally and I don't think that matters for the scenario you are looking at. If you really want to do something like that, look at the ngram analyzer. https://ravendb.net/docs/article-page/6.0/csharp/indexes/using-analyzers#analyzers-that-tokenize-according-to-the-defined-number-of-characters

**Comment by ayende** [comment]
Did you notice:

```
                    .whereEquals("jobTitle", title, false)
```

But the data you are showing has: `job_title`

Fairly certain that this is the problem you run into.

Fuzzy search is something very different.

**Reply by rahul-raj** [reply]
When I replaced `toList()` with `toString()`, then it just gave me query operation as response (no errors) :
```
from 'Jobs' where fuzzy(job_title = $p0, 0.5)
```

I have corrected `jobTitle` to `job_title` matching the field name in the document. 
I was thinking earlier that I should match it with POJO attribute name. 
Thanks for providing more insights here. It seems like `search()` is pretty much what I need, until `vector search` feature comes into `RavenDB` :)



Just one question though. Is there a difference between:

```
try(IDocumentSession session = store.openSession()) {
            List<Job> jobs  = session
                    .advanced()
                    .documentQuery(Job.class)
                    .search("job_title", title)
                    .toList();
            return ResponseEntity.ok(jobs);
        }
```

and

```
try(IDocumentSession session = store.openSession()) {
            List<Job> jobs  = session
                    .query(Job.class)
                    .search("job_title", title)
                    .toList();
            return ResponseEntity.ok(jobs);
        }
```

I'm seeing pretty much the same set of results returned.

**Reply by ml054** [reply]
In Java API there is no difference between session.query and session.advanced.documentQuery. It is alias. 

In c# session.query uses LINQ syntax which is not available in JDK.

**Reply by ml054** [reply]
About casing conventions, as @ayende wrote:
You're using custom naming conventions: `jobTitle` becomes `job_title` in target document. Since queries are executed on server side, you should simply match casing stored on the server. If case of confusion you can simply check target fields using RavenDB Studio. 

Also this view might be handy to easily check what query is send to the server:
http://live-test.ravendb.net/studio/index.html#admin/settings/trafficWatch

I hope it help. :)

**Reply by rahul-raj** [reply]
@ml054 
That's a nice feature! Thanks for sharing with me.
I'm still playing around the features of RavenDB!
